### PR TITLE
Insert into a table with a list of columns

### PR DESCRIPTION
**Pull request by kokosing**
Insert into a table with a list of columns

Now it is possible to specify list of columns with INSERT INTO e.g.
INSERT INTO orders (orderkey) VALUES(1);

When the list of column names is specified the columns produced by the
query must exactly match to them. Each column not being present in
the column list will be filled with a ``null`` value.
In case the list of column names is not given then the columns produced
by query must exactly match the columns in the table being inserted
into.

When the list of columns is present then there is added additional
ProjectNode to logical plan which represents mapping between query
columns and table columns. Also it takes care of table columns not
present in a list.

### CODE REVIEW

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 1]
```diff
@@ -7,18 +7,19 @@ Synopsis
...
-    INSERT INTO table_name query
+    INSERT INTO table_name [ ( column [, ... ] )  query
 
 Description
 -----------
 
 Insert new rows into a table.
 
-.. note::
+When the list of column names is specified the columns produced by the
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]: Comma after "specified"

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 2]
```diff
@@ -7,18 +7,19 @@ Synopsis
...
+    INSERT INTO table_name [ ( column [, ... ] )  query
 
 Description
 -----------
 
 Insert new rows into a table.
 
-.. note::
+When the list of column names is specified the columns produced by the 
+query must exactly match to them. Each column not being present in
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]: Remove "being"

Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]: Reword it slightly:

"Each column in the table not present in the column list ..."

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 3]
```diff
@@ -7,18 +7,19 @@ Synopsis
 
 .. code-block:: none
 
-    INSERT INTO table_name query
+    INSERT INTO table_name [ ( column [, ... ] )  query
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]: Remove the extra space between ")" and "query"

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 4]
```diff
@@ -7,18 +7,19 @@ Synopsis
...
 Description
 -----------
 
 Insert new rows into a table.
 
-.. note::
+When the list of column names is specified the columns produced by the 
+query must exactly match to them. Each column not being present in 
+the column list will be filled with a ``null`` value.
+In case the list of column names is not given then the columns produced
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]: "If the list of column names is not specified, then the columns ..."

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 5]
```diff
@@ -7,18 +7,19 @@ Synopsis
...
 
-    INSERT INTO table_name query
+    INSERT INTO table_name [ ( column [, ... ] )  query
 
 Description
 -----------
 
 Insert new rows into a table.
 
-.. note::
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]: Slightly rephrased text:

```
If the list of column names is specified, they must exactly match the list
of columns produced by the query. Each column in the table not present in the
column list will be filled with a ``null`` value. Otherwise, if the list of
columns is not specified, the columns produced by the query must exactly match
the columns in the table being inserted into.

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 6]
```diff
@@ -35,3 +36,14 @@ Insert a single row into the ``cities`` table::
 Insert multiple rows into the ``cities`` table::
 
     INSERT INTO cities VALUES (2, 'San Jose'), (3, 'Oakland');
+
+Insert a sinle row into the ``nation`` table with usage of the column list::
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]: Typo "sinle" -> "single"

Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]:
```
Insert a single row into the ``nation`` table with the specified column list::

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 7]
```diff
@@ -35,3 +36,14 @@ Insert a single row into the ``cities`` table::
...
+
+    INSERT INTO nation (n_nationkey, n_name, n_regionkey, n_comment)
+    VALUES (26, 'POLAND', 3, 'no comment');
+
+Above insert but with missing ``n_comment`` column in the column list,
```
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]:
```
Insert a row without specifying the ``n_comment`` column. That column will be ``null``::

File: presto-docs/src/main/sphinx/sql/insert.rst (hunk 6 above)
Comment (electrum) [inline presto-docs/src/main/sphinx/sql/insert.rst]: Let's remove all the `_n` prefixes from the example columns, since it makes the examples harder to read.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -438,13 +439,36 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
-                .map(ColumnMetadata::getType)
-                .collect(toImmutableList());
+        List<ColumnMetadata> allTableColumns = metadata.getTableMetadata(session, targetTableHandle.get()).getColumns();
+        final Iterable<Type> tableTypes;
+        if (insert.getColumns().isPresent()) {
+            List<String> insertColumns = insert.getColumns().get();
+            analysis.setInsertColumns(insert.getColumns().get());
 
-        Iterable<Type> queryTypes = transform(descriptor.getVisibleFields(), Field::getType);
+            if (new HashSet<>(insertColumns).size() != insertColumns.size()) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Include the column name in the error message:
```java
Set<String> columnNames = new HashSet<>();
for (String name : insertColumns) {
    if (!columnNames.add(name)) {
        throw new SemanticException(DUPLICATE_COLUMN_NAME, insert, "Insert column name is specified more than once: " + name);
    }
}

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 9]
```diff
@@ -438,13 +439,36 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
+            }
+
+            tableTypes = insertColumns.stream()
+                    .map(columnName -> (ColumnMetadata) allTableColumns.stream()
+                            .filter(column -> column.getName().equalsIgnoreCase(columnName))
+                            .findFirst()
+                            .orElseThrow(() -> new SemanticException(
+                                    MISSING_COLUMN,
+                                    insert,
+                                    "Insert column: '%s' is missing in destination table",
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]:
```java
"Insert column is missing in destination table: %s"

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 10]
```diff
@@ -438,13 +439,36 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
+        if (insert.getColumns().isPresent()) {
+            List<String> insertColumns = insert.getColumns().get();
+            analysis.setInsertColumns(insert.getColumns().get());
 
-        Iterable<Type> queryTypes = transform(descriptor.getVisibleFields(), Field::getType);
+            if (new HashSet<>(insertColumns).size() != insertColumns.size()) {
+                throw new SemanticException(DUPLICATE_COLUMN_NAME, insert, "Insert columns have at least one duplicated column");
+            }
+
+            tableTypes = insertColumns.stream()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: This nested stream code is really hard to read. Let's do it as a normal for-each loop.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 11]
```diff
@@ -128,11 +131,44 @@ private RelationPlan createInsertPlan(Analysis analysis)
...
+
+            Map<Symbol, Expression> assignments = columns.stream()
+                    .filter(column -> !column.isHidden())
+                    .map(column -> symbolAllocator.newSymbol(column.getName(), column.getType()))
+                    .collect(toImmutableMapReversed(tableColumnAsSymbol -> {
+                        int i = insertColumns.get().indexOf(tableColumnAsSymbol.getName());
+                        if (i >= 0) {
+                            return source.getOutputSymbols().get(i).toQualifiedNameReference();
+                        }
+                        else {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: The `else` is redundant

File: presto-docs/src/main/sphinx/sql/insert.rst (hunk 5 above)
Comment (kokosing) [inline presto-docs/src/main/sphinx/sql/insert.rst]: Thanks. It sounds much better when it is written by native speaker. Next time I will try harder to write something better.

**Comment by kokosing** [comment]
Comments addressed.

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 12]
```diff
@@ -7,18 +7,19 @@ Synopsis
 
 .. code-block:: none
 
-    INSERT INTO table_name query
+    INSERT INTO table_name [ ( column [, ... ] ) query
```
Comment (losipiuk) [inline presto-docs/src/main/sphinx/sql/insert.rst]: missing closing `]`

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 13]
```diff
@@ -35,3 +36,14 @@ Insert a single row into the ``cities`` table::
 Insert multiple rows into the ``cities`` table::
 
     INSERT INTO cities VALUES (2, 'San Jose'), (3, 'Oakland');
+
+Insert a sinle row into the ``nation`` table with the specified column list::
```
Comment (losipiuk) [inline presto-docs/src/main/sphinx/sql/insert.rst]: a/sinle/single/

File: presto-docs/src/main/sphinx/sql/insert.rst [hunk 14]
```diff
@@ -35,3 +36,14 @@ Insert a single row into the ``cities`` table::
...
+
+    INSERT INTO nation (nationkey, name, regionkey, comment)
+    VALUES (26, 'POLAND', 3, 'no comment');
```
Comment (losipiuk) [inline presto-docs/src/main/sphinx/sql/insert.rst]: s/no comment/best country ever/ ;)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 15]
```diff
@@ -438,13 +439,40 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
+        final Iterable<Type> tableTypes;
+        if (insert.getColumns().isPresent()) {
+            List<String> insertColumns = insert.getColumns().get();
+            analysis.setInsertColumns(insert.getColumns().get());
 
-        Iterable<Type> queryTypes = transform(descriptor.getVisibleFields(), Field::getType);
+            Set<String> columnNames = new HashSet<>();
+            for (String name : insertColumns) {
+                if (!columnNames.add(name)) {
+                    throw new SemanticException(DUPLICATE_COLUMN_NAME, insert, "Insert column name is specified more than once: %s", name);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: maybe "Insert column name '%s' is specified more than once"

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 16]
```diff
@@ -438,13 +439,40 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
+                }
+            }
+
+            ImmutableList.Builder<Type> tableTypesBuilder = ImmutableList.builder();
+            for (String insertColumn : insertColumns) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Use collector

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 17]
```diff
@@ -438,13 +439,40 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
+                ColumnMetadata column = allTableColumns.stream()
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Maybe extract checkColumnExists?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 18]
```diff
@@ -438,13 +439,40 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
+                        .filter(tableColumn -> tableColumn.getName().equalsIgnoreCase(insertColumn))
+                        .findFirst()
+                        .orElseThrow(() -> new SemanticException(
+                                MISSING_COLUMN,
+                                insert,
+                                "Insert column is missing in destination table: %s", insertColumn));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: "Insert column '%s' is missing in destination table"

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 19]
```diff
@@ -128,11 +131,42 @@ private RelationPlan createInsertPlan(Analysis analysis)
...
+        List<ColumnMetadata> columns = getVisibleColumns(tableMetadata);
+
+        RelationPlan plan = createRelationPlan(analysis);
+
+        Optional<List<String>> insertColumns = analysis.getInsertColumns();
+        if (insertColumns.isPresent() && !insertColumns.get().equals(columnNames)) {
+            PlanNode source = plan.getRoot();
+
+            Map<Symbol, Expression> assignments = columns.stream()
+                    .filter(column -> !column.isHidden())
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: this .filter is not needed (you have only visible columns on `columns` variable)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 20]
```diff
@@ -438,13 +439,40 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
         analysis.setInsertTarget(targetTableHandle.get());
 
-        List<ColumnMetadata> columns = metadata.getTableMetadata(session, targetTableHandle.get()).getColumns();
-        Iterable<Type> tableTypes = columns.stream()
-                .filter(column -> !column.isHidden())
-                .map(ColumnMetadata::getType)
-                .collect(toImmutableList());
+        List<ColumnMetadata> allTableColumns = metadata.getTableMetadata(session, targetTableHandle.get()).getColumns();
+        final Iterable<Type> tableTypes;
+        if (insert.getColumns().isPresent()) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: maybe, instead having `if/else` block, always execute code for insertColumns present.
But when those are not present treat them as all visible columns from target table.
```
List<String> insertColumns = analysis.getInsertColumns().orElse(allTableColumnsNames)
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 21]
```diff
@@ -438,13 +439,40 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
+            for (String name : insertColumns) {
+                if (!columnNames.add(name)) {
+                    throw new SemanticException(DUPLICATE_COLUMN_NAME, insert, "Insert column name is specified more than once: %s", name);
+                }
+            }
+
+            ImmutableList.Builder<Type> tableTypesBuilder = ImmutableList.builder();
+            for (String insertColumn : insertColumns) {
+                ColumnMetadata column = allTableColumns.stream()
+                        .filter(tableColumn -> tableColumn.getName().equalsIgnoreCase(insertColumn))
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: you should disallow hidden columns here

**Comment by losipiuk** [comment]
reviewed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 15 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: It used be this way. I changed it @electrum request.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 16 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: It used be this way. I changed it @electrum request.

The code was difficult to read this way.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 15 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I think it is better to put the user string on the right side of the error when possible. This avoid issues with quoting, spaces, etc. that can appear in identifiers (which we don't handle at all now in error messages).

**Comment by cberner** [comment]
It looks like the tests are failing. Can you check Travis?

**Comment by kokosing** [comment]
@cberner Sorry about this. It was simple mistake during refactoring. It should be fixed now.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 22]
```diff
@@ -113,13 +116,13 @@ private RelationPlan createTableCreationPlan(Analysis analysis)
 
         TableMetadata tableMetadata = createTableMetadata(destination, getOutputTableColumns(plan), analysis.getCreateTableProperties(), plan.getSampleWeight().isPresent());
         if (plan.getSampleWeight().isPresent() && !metadata.canCreateSampledTables(session, destination.getCatalogName())) {
-          throw new PrestoException(NOT_SUPPORTED, "Cannot write sampled data to a store that doesn't support sampling");
+            throw new PrestoException(NOT_SUPPORTED, "Cannot write sampled data to a store that doesn't support sampling");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: unrelated formatting change

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 23]
```diff
@@ -438,13 +440,34 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
...
-                .collect(toImmutableList());
+        TableMetadata tableMetadata = metadata.getTableMetadata(session, targetTableHandle.get());
+        List<ColumnMetadata> allTableColumns = tableMetadata.getColumns();
 
-        Iterable<Type> queryTypes = transform(descriptor.getVisibleFields(), Field::getType);
+        List<String> insertColumns = insert.getColumns().orElse(tableMetadata.getVisibleColumnNames());
+        analysis.setInsertColumns(insertColumns);
+
+        Set<String> columnNames = new HashSet<>();
+        for (String name : insertColumns) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: All these validations shouldn't be done if the columns were not listed explicitly. I'd restructure it as:

```java
List<String> insertColumns = tableMetadata.getVisibleColumnNames();
if (insert.getColumns().isPresent()) { 
   ... validate duplicate names
   ... ensure columns exist in target table
}

analysis.setInsertColumns(insertColumns);
... derive types ...
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 24]
```diff
@@ -128,11 +131,41 @@ private RelationPlan createInsertPlan(Analysis analysis)
 
         TableMetadata tableMetadata = metadata.getTableMetadata(session, target);
 
+        List<String> visibleTableColumnNames = tableMetadata.getVisibleColumnNames();
+        List<ColumnMetadata> visibleTableColumns = tableMetadata.getVisibleColumns();
+
+        RelationPlan plan = createRelationPlan(analysis);
+
+        Optional<List<String>> insertColumns = analysis.getInsertColumns();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: Why would `analysis.getInsertColumns()` ever return "missing"? The analyzer always sets it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 25]
```diff
@@ -128,11 +131,41 @@ private RelationPlan createInsertPlan(Analysis analysis)
...
+        if (insertColumns.isPresent() && !insertColumns.get().equals(visibleTableColumnNames)) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: You should be able to do this unconditionally. The optimizer should prune any redundant projections.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 26]
```diff
@@ -128,11 +131,41 @@ private RelationPlan createInsertPlan(Analysis analysis)
...
+            PlanNode source = plan.getRoot();
+
+            Map<Symbol, Expression> assignments = visibleTableColumns.stream()
+                    .map(column -> symbolAllocator.newSymbol(column.getName(), column.getType()))
+                    .collect(toImmutableMapReversed(tableColumnAsSymbol -> {
+                        int i = insertColumns.get().indexOf(tableColumnAsSymbol.getName());
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: This is incorrect. The names created by SymbolAllocator won't necessarily match the original column names. You should think of them as "variables" that reference some input.

Also, `indexOf` may break if the case of the value in insertColumns doesn't match the case of the name in the metadata. Instead, you should turn names into ColumnHandles in the analyzer and record those in the Analysis.

In general, I think what you want here is something like:

```java
List<ColumnHandle> insertColumns = analysis.getInsertColumns();

TableMetadata tableMetadata = metadata.getTableMetadata(session, target);
Map<String, ColumnHandle> columns = metadata.getColumnHandles(session, target);

ImmutableMap.Builder<Symbol, Expression> assignments = ImmutableMap.builder();
for (ColumnMetadata column :  tableMetadata.getVisibleColumns()) {
    Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
    int index = insertColumns.indexOf(columns.get(column.getName()));
    if (index < 0) {
        assignments.put(output, new NullLiteral());
    }
    else {
        assignments.put(output, plan.getSymbol(index).toQualifiedNameReference());
    }
}

```

File: presto-main/src/main/java/com/facebook/presto/util/ImmutableCollectors.java [hunk 27]
```diff
@@ -74,4 +74,16 @@ private ImmutableCollectors() {}
                 },
                 ImmutableMap.Builder::build);
     }
+
+    public static <K, V> Collector<K, ?, ImmutableMap<K, V>> toImmutableMapReversed(Function<K, V> keyMapper)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/ImmutableCollectors.java]: The name of this method is weird and confusing. Also, I don't think you need the method at all (see my comment above on how to structure the code in the LogicalPlanner)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 23 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java (hunk 22 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java (hunk 24 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java (hunk 25 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java (hunk 26 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: Done

File: presto-main/src/main/java/com/facebook/presto/util/ImmutableCollectors.java (hunk 27 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/ImmutableCollectors.java]: done

**Comment by kokosing** [comment]
@martint comments applied and rebased

**Comment by martint** [comment]
Unquoted identifiers are expected to be case-insensitive. 

Presto currently doesn't really handle quoted identifiers properly -- it handles them from a syntactic point of view, but doesn't treat them correctly. Yes, this is something we're planning to fix.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 28]
```diff
@@ -279,12 +279,16 @@ public void testInsert()
...
         assertQuery("INSERT INTO test_insert " + query, "SELECT count(*) FROM orders");
 
-        assertQuery("SELECT * FROM test_insert", query + " UNION ALL " + query);
+        assertQuery("SELECT * FROM test_insert", query);
+
+        assertQuery("INSERT INTO test_insert (orderkey) VALUES (-1)", "SELECT 1");
+
+        assertQuery("SELECT * FROM test_insert", query + " UNION ALL SELECT null, -1");
 
         assertQueryTrue("DROP TABLE test_insert");
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Can you add tests that insert into multiple columns in different orders?

```
INSERT INTO xxx (a, b) VALUES ....
INSERT INTO xxx (b, a) VALUES ....
```

**Comment by martint** [comment]
Other than that last request, this looks good. Please rebase, since there have been some changes to the analyzer recently and I'll merge it.

**Comment by kokosing** [comment]
Thanks @martint, rebased and tests were added.

**Comment by martint** [comment]
Merged, thanks!
